### PR TITLE
Add takeOrderOnKyber

### PR DESCRIPTION
**Pull request by luongnt95**

### CODE REVIEW

File: src/contracts/fund/trading/transactions/takeOrderOnKyber.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+} from '~/utils/solidity/transactionFactory';
+import { getHub } from '../../hub/calls/getHub';
+import { getRoutes } from '../../hub/calls/getRoutes';
+import { ensureSufficientBalance } from '~/contracts/dependencies/token/guards/ensureSufficientBalance';
+import { ensureFundOwner } from '~/contracts/fund/trading/guards/ensureFundOwner';
+import { ensureIsNotShutDown } from '~/contracts/fund/hub/guards/ensureIsNotShutDown';
+import { ensureKyberTakePermitted } from '../guards/ensureKyberTakePermitted';
+import { FunctionSignatures } from '../utils/FunctionSignatures';
+
+// The order needs to be signed by the manager
```
Comment (jennazenk) [inline src/contracts/fund/trading/transactions/takeOrderOnKyber.ts]: can probably remove that comment

File: src/contracts/fund/trading/transactions/takeOrderOnKyber.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+const prepareArgs: PrepareArgsFunction<TakeOrderOnKyberArgs> = async (
+  environment,
+  { makerQuantity, takerQuantity, fillTakerQuantity = takerQuantity },
+  contractAddress,
+) => {
+  const exchangeIndex = await getExchangeIndex(environment, contractAddress, {
+    exchange: Exchanges.KyberNetwork,
+  });
+
+  const takerAddress = takerQuantity.token.address;
```
Comment (jennazenk) [inline src/contracts/fund/trading/transactions/takeOrderOnKyber.ts]: think this should `takerAddress` should be `takerAsset`  and `makerAddress` should be `makerAsset`

File: src/contracts/fund/trading/calls/isKyberTakePermitted.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import web3EthAbi from 'web3-eth-abi';
+import { getContract } from '~/utils/solidity/getContract';
+import { Contracts, Exchanges } from '~/Contracts';
+import { getRoutes } from '~/contracts/fund/hub/calls/getRoutes';
+import { getHub } from '~/contracts/fund/hub/calls/getHub';
+import { FunctionSignatures } from '../utils/FunctionSignatures';
+import { Environment } from '~/utils/environment/Environment';
+import { Address, QuantityInterface } from '@melonproject/token-math';
+
+const isKyberTakePermitted = async (
```
Comment (schmidsi) [inline src/contracts/fund/trading/calls/isKyberTakePermitted.ts]: I'm not 100% sure, but this looks to me like a "duplicated code" smell. What is the actual difference between `isKyberTakePermitted` and `isOasisDexMakePermitted`. I only see fillTakerQuantity which could be defaulted to `0`.

So better to refactor this function to `isTakePermitted` that works for all exchanges. Like the contracts are implemented if I understand them right.

https://github.com/melonproject/protocol/blob/develop/src/contracts/fund/policies/risk-management/PriceTolerance.sol#L73:L79

File: src/contracts/fund/trading/guards/ensureKyberTakePermitted.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { ensure } from '~/utils/guards/ensure';
+import { isKyberTakePermitted } from '../calls/isKyberTakePermitted';
+import { Environment } from '~/utils/environment/Environment';
+import { Address, QuantityInterface } from '@melonproject/token-math';
+
+const ensureKyberTakePermitted = async (
```
Comment (schmidsi) [inline src/contracts/fund/trading/guards/ensureKyberTakePermitted.ts]: Same here. Could be just: `ensureTakePermitted`

File: src/contracts/fund/trading/calls/isKyberTakePermitted.ts (hunk 3 above)
Comment (luongnt95) [reply src/contracts/fund/trading/calls/isKyberTakePermitted.ts]: https://github.com/melonproject/protocol/blob/develop/src/contracts/fund/trading/calls/isOasisDexTakePermitted.ts#L47

Comment (schmidsi) [reply src/contracts/fund/trading/calls/isKyberTakePermitted.ts]: Right. Please use the Exchange enum here: https://github.com/melonproject/protocol/blob/develop/src/Contracts.ts#L269

**Pull request merged**